Add docs for VCS Speculative Plan Management (#749)
hashicorp/terraform-docs-common: website/docs/cloud-docs/users-teams-organizations/organizations/vcs-speculative-plan-management.mdx

---
page_title: Automatically cancel plan-only runs triggered by outdated commits
description: Learn how to configure HCP Terraform to automatically cancel Terraform plan operations triggered by pull requests when new commits are pushed to the VCS.
---

# Automatically cancel plan-only runs

This topic describes how to configure HCP Terrraform to automatically cancel plan-only Terraform run triggered by pull requests in the VCS. 

## Introduction
 

> When connected to a VCS, HCP Terraform can automatically start a Terraform run that performs a `terraform plan` operation when someone creates a pull request (PR) in the repository. Refer to [Connecting to VCS](/terraform/cloud-docs/vcs) for additional information. 

Reviewer: ```suggestion
When connected to a VCS, HCP Terraform can automatically start a Terraform run that performs a `terraform plan` operation when someone creates a pull request (PR) in the repository. Refer to [Connecting to VCS](/terraform/cloud-docs/vcs) for additional information. 

```

> When team members push new commits to the same branch, HCP Terraform starts new run that performs a `terraform plan` operation. But as team members push new commits, the queue of Terraform runs can cause delays and reduce efficiency. 

Reviewer: ```suggestion
When team members push new commits to the same branch, HCP Terraform starts new run that performs a `terraform plan` operation. But as team members push new commits, the queue of Terraform runs can cause delays and reduce efficiency. 

```

---
page_title: VCS Speculative Plan Management
description: Learn how VCS Speculative Plans Management work in HCP Terraform.
---

# VCS Speculative Plan Management

VCS Speculative Plan Management setting automatically cancel's unfinished speculative plans triggered by pull requests when a new commit is pushed in VCS workflows.

## Managing Organization Speculative Plan Settings 

To view and manage an organization’s Speculative Plan settings, click Settings, followed by General under Version Control, then view the Manage speculative plans part of the page.

Once the organization setting is enabled, HCP Terraform will cancel ongoing or pending speculative plans if new commits are received on the same branch.

![Screenshot: VCS Speculative Plan Management Settings](/img/docs/vcs-speculative-plan-management-general-settings.png)

> ## Run Details Page
> 
> When a newer commit gets pushed to the same branch, a user can see the message mentioning that the run was automatically canceled.
> 
> ![Screenshot: VCS Speculative Plan Management Run Details](/img/docs/vcs-speculative-plan-management-run-details.png)
> 
> ## Updated VCS Status Checks
> 
> For non-aggregated status checks, we now show a different message for the auto canceled plan than the normally canceled one (which would show “Terraform plan canceled”).
> 
> ![Screenshot: VCS Speculative Plan Management Non Aggregated Status Checks](/img/docs/vcs-speculative-plan-management-non-aggregated-status-checks.png)
> 
> For aggregated status check, we now include the count of automatically canceled runs in the aggregated result, separating it from the count of manually canceled ones.
> 
> ![Screenshot: VCS Speculative Plan Management Auto Cancelled Run](/img/docs/vcs-speculative-plan-management-auto-cancelled-run.png)
> 
> Also, in the aggregated status page, we now include the message that the run has been auto canceled, using the same message that we would send for non-aggregated status checks.
> 
> ![Screenshot: VCS Speculative Plan Management Auto Cancelled Run Message](/img/docs/vcs-speculative-plan-management-auto-cancelled-run-message.png)
> 
> ## Aggregated Status Page
> 
> When more than one run is automatically canceled for a commit, we’ll show an alert text below the "Resources to be changed" section, as a reminder that the section may not reflect a complete result if all the runs of the commit were to reach completion.
> 
> ![Screenshot: VCS Speculative Plan Management Aggregated Status Page](/img/docs/vcs-speculative-plan-management-aggregated-status.png)

Reviewer: ```suggestion

## Automated cancellation notifications

When the **Automatically cancel speculative plans for outdated commits** option is enabled, HCP Terraform notifies you about plan-only runs that are canceled as a result of the setting.  Notifications appear in the following screens:

-  **Run details page**. Refer to [Viewing and Managing Runs](/terraform/cloud-docs/run/manage) for additional information.

- **VCS status checks**. When the **Non-aggregated status checks** option is enabled in the version control settings, the notification explicitly states when a plan has been canceled automatically.  

    When the **Aggregated status checks** option is enabled, HCP Terraform includes canceled plans in the result and identifies them separately from manually canceled plans. 
     
- **Aggregated status page**. HCP Terraform prints the cancelation message in the aggregated status page in the  **Resources to be changed** section. The section may not reflect a complete result if all runs associated with the commit reach completion.
```
Removed screenshots. 
Updated tense, voice, and text formatting per the style guide.
 
I could not find relevant pages to link to for each screen we mention. Can you add links similar to the run details bullet?

Author: @trujillo-adam Thanks for reviewing the PR, I have incorporate all the suggestions! 

> I also wonder if we should really create an entirely new topic for this workflow. To me, it would make more sense to describe enabling the option as a step in the instructions for starting a run from commits: https://developer.hashicorp.com/terraform/cloud-docs/run/ui

As you suggested [this](https://developer.hashicorp.com/terraform/cloud-docs/run/ui#speculative-plans-on-pull-requests) might be a better place. I'll discuss with my team and update it there.

Author: We think it might be helpful to create a new topic for this workflow alongside the [vcs-status-checks](https://developer.hashicorp.com/terraform/cloud-docs/users-teams-organizations/organizations/vcs-status-checks) as they both are closely related and more intuitive for users. However I have references in both places:
1. Org settings page
2. Run UI page
